File: src/com/polaris/engine/gui/GuiScreen.java
package com.polaris.engine.gui;

import com.polaris.engine.App;
import com.polaris.engine.gui.content.GuiContent;
import com.polaris.engine.options.Settings;
import com.polaris.engine.util.MathHelper;
import org.lwjgl.glfw.GLFW;

import java.util.ArrayList;
import java.util.List;

import static org.lwjgl.glfw.GLFW.GLFW_KEY_ESCAPE;

public abstract class GuiScreen<T extends Settings> extends Gui<T>
{
	
	protected GuiContent<T> focusedElement;
	private List<GuiContent<T>> elementList;
	private int elementCounter = 0;
	
	public GuiScreen(App<T> app)
	{
		super(app);
		
		elementList = new ArrayList<GuiContent<T>>();
	}
	
	public GuiScreen(GuiScreen<T> gui)
	{
		super(gui.application, gui, 0);
		
		elementList = new ArrayList<GuiContent<T>>();
	}
	
	public void update(double delta)
	{
		super.update(delta);
		
		double mouseX = input.getMouseX();
		double mouseY = input.getMouseY();
		
		if (input.getKey(GLFW_KEY_ESCAPE).isPressed())
		{
			if (input.getKey(GLFW.GLFW_KEY_LEFT_SHIFT).isPressed())
			{
				gameSettings.changeWindowMode();
			}
			else
			{
				if (this.getParent() != null)
				{
					application.reinitGui(this.getParent());
				}
				else
				{
					application.close();
				}
			}
		}
		
		for (GuiContent<T> element : elementList)
		{
			if (MathHelper.inBounds(mouseX, mouseY, element.getBounds()))
			{
				//flag = element.handleInput();
			}
			element.update(delta);
		}
	}
	
	public void render(double delta)
	{
		super.render(delta);
		for (GuiContent<T> element : elementList)
		{
			element.render(delta);
		}
	}
	
	public void close()
	{
		this.focusedElement = null;
	}
	
	public boolean mouseClick(int mouseId)
	{
		for (GuiContent<T> element : elementList)
		{
			if (MathHelper.inBounds(input.getMouseX(), input.getMouseY(), element.getBounds()))
			{
				boolean flag = element.nMouseClick(mouseId);
				if (flag && element != focusedElement)
				{
					unbindCurrentElement(element);
				}
				return flag;
			}
		}
		unbindCurrentElement();
		return false;
	}
	
	public void unbindCurrentElement(GuiContent<T> e)
	{
		unbindCurrentElement();
		focusedElement = e;
	}
	
	public void unbindCurrentElement()
	{
		if (focusedElement != null)
		{
			focusedElement.unbind();
			focusedElement = null;
		}
	}
	
	public void mouseHeld(int mouseId)
	{
		if (focusedElement != null && focusedElement.nMouseHeld(mouseId))
		{
			unbindCurrentElement();
		}
	}
	
	public void mouseRelease(int mouseId)
	{
		if (focusedElement != null && !focusedElement.nMouseRelease(mouseId))
		{
			unbindCurrentElement();
		}
	}
	
	public void mouseScroll(double xOffset, double yOffset)
	{
		if (focusedElement != null && focusedElement.nMouseScroll(xOffset, yOffset))
		{
			unbindCurrentElement();
		}
	}
	
	public int keyPressed(int keyId, int mods)
	{
		if (focusedElement != null)
		{
			return focusedElement.nKeyPressed(keyId, mods);
		}
		if (keyId == GLFW_KEY_ESCAPE)
		{
			if ((mods & 1) == 1)
			{
				gameSettings.changeWindowMode();
			}
			else
			{
				if (getParent() != null)
				{
					getParent().reinit();
					application.reinitGui(getParent());
					return 0;
				}
				else
				{
					application.close();
				}
			}
		}
		return -1;
	}
	
	public int keyHeld(int keyId, int called, int mods)
	{
		if (focusedElement != null)
		{
			return focusedElement.nKeyHeld(keyId, called, mods);
		}
		return -1;
	}
	
	public void keyRelease(int keyId, int mods)
	{
		if (focusedElement != null && focusedElement.nKeyRelease(keyId, mods))
		{
			unbindCurrentElement();
		}
	}
	
	public void addElement(GuiContent<T> e)
	{
		e.init(this, elementCounter);
		elementCounter++;
		elementList.add(e);
	}
	
	public void removeElement(int i)
	{
		elementList.remove(i).close();
	}
	
	public void removeElements(int i, int i1)
	{
		for (int j = i1 - 1; j >= i; j--)
		{
			elementList.remove(j).close();
		}
	}
	
	public int getSize()
	{
		return elementList.size();
	}
	
	public void elementUpdate(GuiContent<T> e, int actionId)
	{
	}
	
	public void clearElements()
	{
		elementList.clear();
	}
	
	public GuiContent<T> getCurrentElement()
	{
		return focusedElement;
	}
	
	public void setCurrentElement(int id)
	{
		focusedElement = this.getElement(id);
	}
	
	public GuiContent<T> getElement(int i)
	{
		return elementList.get(i);
	}
	
}

File: src/polaris/okapi/gui/GuiScreen.kt
package polaris.okapi.gui

import polaris.okapi.App
import polaris.okapi.gui.content.GuiContent
import polaris.okapi.options.WindowMode
import polaris.okapi.util.inBounds
import org.lwjgl.glfw.GLFW.GLFW_KEY_ESCAPE
import java.util.ArrayList

/**
 * Created by Killian Le Clainche on 12/10/2017.
 */

abstract class GuiScreen : Gui {

    private var elementList: MutableList<GuiContent> = ArrayList()
    private var elementCounter = 0

    var focusedElement: GuiContent? = null
        protected set

    val size: Int
        get() = elementList.size

    constructor(app: App) : super(app)
    constructor(gui: GuiScreen) : super(gui.application, gui, 0.0)

    override fun render(delta: Double) {
        super.render(delta)

        val mouseX = settings.mouse.x
        val mouseY = settings.mouse.y

        for (element in elementList) {
            if (inBounds(mouseX, mouseY, element.bounds)) {
                //flag = element.handleInput();
            }
            element.update(delta)
        }

        for (element in elementList) {
            element.render(delta)
        }
    }

    override fun close() {
        this.focusedElement = null
    }

    fun mouseClick(mouseId: Int): Boolean {
        for (element in elementList) {
            if (inBounds(settings.mouse.x, settings.mouse.y, element.bounds)) {
                val flag = element.nMouseClick(mouseId)
                if (flag && element !== focusedElement) {
                    unbindCurrentElement(element)
                }
                return flag
            }
        }
        unbindCurrentElement()
        return false
    }

    fun unbindCurrentElement(e: GuiContent) {
        unbindCurrentElement()
        focusedElement = e
    }

    fun unbindCurrentElement() {
        if (focusedElement != null) {
            focusedElement!!.unbind()
            focusedElement = null
        }
    }

    fun mouseHeld(mouseId: Int) {
        if (focusedElement != null && focusedElement!!.nMouseHeld(mouseId)) {
            unbindCurrentElement()
        }
    }

    fun mouseRelease(mouseId: Int) {
        if (focusedElement != null && !focusedElement!!.nMouseRelease(mouseId)) {
            unbindCurrentElement()
        }
    }

    fun mouseScroll(xOffset: Double, yOffset: Double) {
        if (focusedElement != null && focusedElement!!.nMouseScroll(xOffset, yOffset)) {
            unbindCurrentElement()
        }
    }

    fun keyPressed(keyId: Int, mods: Int): Int {
        if (focusedElement != null) {
            return focusedElement!!.nKeyPressed(keyId, mods)
        }
        if (keyId == GLFW_KEY_ESCAPE) {
            if (mods and 1 == 1) {
                settings.windowMode = WindowMode.WINDOWED
            } else {
                if (parent != null) {
                    parent.reinit()
                    application.reinitGui(parent)
                    return 0
                } else {
                    application.isRunning = false
                }
            }
        }
        return -1
    }

    fun keyHeld(keyId: Int, called: Int, mods: Int): Int {
        return if (focusedElement != null) {
            focusedElement!!.nKeyHeld(keyId, called, mods)
        } else -1
    }

    fun keyRelease(keyId: Int, mods: Int) {
        if (focusedElement != null && focusedElement!!.nKeyRelease(keyId, mods)) {
            unbindCurrentElement()
        }
    }

    fun addElement(e: GuiContent) {
        e.init(this, elementCounter)
        elementCounter++
        elementList.add(e)
    }

    fun removeElement(i: Int) {
        elementList.removeAt(i).close()
    }

    fun removeElements(i: Int, i1: Int) {
        for (j in i1 - 1 downTo i) {
            elementList.removeAt(j).close()
        }
    }

    fun elementUpdate(e: GuiContent, actionId: Int) {}

    fun clearElements() {
        elementList.clear()
    }

    fun setCurrentElement(id: Int) {
        focusedElement = this.getElement(id)
    }

    fun getElement(i: Int): GuiContent {
        return elementList[i]
    }

}